pathology/hovernet/infer_torch.py

```diff
@@ -0,0 +1,256 @@
+    GenerateInstanceContour,
+    GenerateInstanceCentroid,
+    GenerateInstanceType,
+)
+
+import os
+import glob
+import torch
+import numpy as np
+from argparse import ArgumentParser
+from monai.data import DataLoader, decollate_batch, Dataset, CacheDataset
+from monai.metrics import DiceMetric
+from monai.metrics.confusion_matrix import ConfusionMatrixMetric
+from monai.losses import DiceLoss
+# from monai.networks.nets import HoVerNet
+from monai.transforms import (
+    Activations,
+    AsDiscrete,
+    AsDiscreted,
+    Compose,
+    ScaleIntensityRanged,
+    CastToTyped,
+    Lambdad,
+    SplitDimd,
+    EnsureChannelFirstd,
+    ComputeHoVerMapsd,
+    RandFlipd,
+    RandRotate90d,
+    RandGaussianSmoothd,
+    GaussianSmooth,
+    FillHoles,
+    BoundingRect,
+    CenterSpatialCropd,
+    SaveImage,
+)
+# from monai.apps.pathology.transforms.post import (
+#     GenerateInstanceContour,
+#     GenerateInstanceCentroid,
+#     GenerateInstanceType
+# )
+
+from monai.utils import set_determinism, convert_to_tensor
+from monai.utils.enums import HoVerNetBranch
+from monai.visualize import plot_2d_or_3d_image
+from skimage import measure
+
+
+def prepare_data(data_dir, phase):
+    data_dir = os.path.join(data_dir, phase)
+
+    files = sorted(
+        glob.glob(os.path.join(data_dir, "*/*.npy")))
+
+    images, labels, inst_maps, type_maps = [], [], [], []
+    for file in files:
+        data = np.load(file)
+        images.append(data[..., :3].transpose(2, 0, 1))
+        inst_maps.append(measure.label(data[..., 3][None]).astype(int))
+        type_maps.append(data[..., 4][None])
+        labels.append(np.array(data[..., 3][None] > 0, dtype=int))
+
+    data_dicts = [
+        {"image": _image, "label": _label, "label_inst": _inst_map, "label_type": _type_map}
+        for _image, _label, _inst_map, _type_map in zip(images, labels, inst_maps, type_maps)
+    ]
+
+    return data_dicts
+
+
+def _dice_info(true, pred, label):
+    true = np.array(true == label, np.int32)
+    pred = np.array(pred == label, np.int32)
+    inter = (pred * true).sum()
+    total = (pred + true).sum()
+    return inter, total
+
+
+def post_process_WS(output, device, return_binary=True, return_centroids=False, output_classes=None):
+    post_trans_seg = Compose([
+        GenerateWatershedMaskd(keys=HoVerNetBranch.NP.value, softmax=True),
+        GenerateInstanceBorderd(keys='mask', hover_map_key=HoVerNetBranch.HV, kernel_size=21),
+        GenerateDistanceMapd(keys='mask', border_key='border', smooth_fn=GaussianSmooth()),
+        GenerateWatershedMarkersd(keys='mask', border_key='border', threshold=0.99, radius=3, postprocess_fn=FillHoles(connectivity=2)),
+        Watershedd(keys='dist', mask_key='mask', markers_key='markers')
+    ])
+    if HoVerNetBranch.NC.value in output.keys():
+        type_pred = Activations(softmax=True)(output[HoVerNetBranch.NC.value])
+        type_pred = AsDiscrete(argmax=True)(output[HoVerNetBranch.NC.value])
```

Comment:
`type_pred = AsDiscrete(argmax=True)(type_pred)`

Reply:
Thanks, sorry for I have made a mistake here.